### PR TITLE
feat: mls migration

### PR DESCRIPTION
**Pull request by PatrykBuniX**
Feature branch for MLS migration feature, for more details see subpages of https://wearezeta.atlassian.net/wiki/spaces/CORE/pages/746488003/Proteus+to+MLS+Migration

### CODE REVIEW

File: src/script/components/MessagesList/Message/ProtocolUpdateMessage/ProtocolUpdateMessage.tsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+          <SystemMessageBase key={message.caption} icon={<Icon.Info />} message={message} />
+        ))}
+      </>
+    );
+  }
+
+  const migratedToMLSMessage = createSystemMessage(
+    t('conversationProtocolUpdatedToMLS', {}, replaceLink(Config.getConfig().URL.SUPPORT.MLS_LEARN_MORE)),
+  );
+  return <SystemMessageBase message={migratedToMLSMessage} icon={<Icon.Info />} />;
```
Comment (atomrc) [inline src/script/components/MessagesList/Message/ProtocolUpdateMessage/ProtocolUpdateMessage.tsx:59]: I'm a little worried about this cyclic dependency (`ProtocolUpdateMessage` depends on `SystemMessageBase` and `SystemMessageBase` depends on `ProtocolUpdateMessage`). 
Also we could somehow fall into an infinite loop if something is messed up.
There might be some room for improvements there

File: src/script/components/MessagesList/Message/SystemMessage/SystemMessageBase.tsx [hunk 2]
```diff
@@ -39,7 +39,9 @@ export const SystemMessageBase: React.FC<SystemMessageProps> = ({message, isSend
       <p className="message-header-label">
         <span className="message-header-label__multiline">
           {isSenderNameVisible && <span className="message-header-sender-name">{unsafeSenderName}</span>}
-          {message.caption && <span className="ellipsis">{message.caption}</span>}
+          {message.caption && (
+            <span className="system-message-caption ellipsis" dangerouslySetInnerHTML={{__html: message.caption}} />
+          )}
```
Comment (atomrc) [inline src/script/components/MessagesList/Message/SystemMessage/SystemMessageBase.tsx:44]: **out of scope**: This is probably not for this PR, but in the future, we should try to move this caption for the `entity` to the `react component` layer and render it a pure React node (not `innerHTML`)

**Review by atomrc — CHANGES_REQUESTED** [review]

File: src/script/conversation/ConversationRepository.ts [hunk 3]
```diff
@@ -1004,6 +1008,28 @@ export class ConversationRepository {
     }
   }
 
+  /**
+   * Get all the group conversations owned by self user's team from the local state.
+   */
+  public readonly getAllSelfTeamOwnedGroupConversations = (): Conversation[] => {
```
Comment (atomrc) [inline src/script/conversation/ConversationRepository.ts]: **suggestion**: the name seems a little long. How about a small simplification?

```suggestion
  public readonly getAllTeamGroupConversations = (): Conversation[] => {
```

File: src/script/conversation/ConversationRepository.ts [hunk 4]
```diff
@@ -1375,7 +1401,7 @@ export class ConversationRepository {
    * Proteus 1:1 conversation will be hidden in the UI and replaced with mls 1:1 conversation.
    *
    * @param proteusConversation - proteus 1:1 conversation
-   * @param mlsConversation - mls 1:1 conversation
+   * @param mlsConversation - mls 1:1 conversation``
```
Comment (atomrc) [inline src/script/conversation/ConversationRepository.ts]:
```suggestion
   * @param mlsConversation - mls 1:1 conversation
```

maybe? 🤷

Comment (przemvs) [inline src/script/conversation/ConversationRepository.ts]: This change is necessary ? :)

File: src/script/conversation/ConversationRepository.ts [hunk 5]
```diff
@@ -2855,6 +2967,8 @@ export class ConversationRepository {
       case ClientEvent.CONVERSATION.LEGAL_HOLD_UPDATE:
       case ClientEvent.CONVERSATION.LOCATION:
       case ClientEvent.CONVERSATION.MISSED_MESSAGES:
+      case ClientEvent.CONVERSATION.JOINED_AFTER_MLS_MIGRATION_FINALISATION:
+      case ClientEvent.CONVERSATION.MLS_MIGRATION_FINALISATION_ONGOING_CALL:
```
Comment (atomrc) [inline src/script/conversation/ConversationRepository.ts]: **suggestion**: I feel the `FINALISATION` here is kind of redundant. WDYT?

```suggestion
      case ClientEvent.CONVERSATION.JOINED_AFTER_MLS_MIGRATION:
      case ClientEvent.CONVERSATION.MLS_MIGRATION_ONGOING_CALL:
```

File: src/script/conversation/MessageRepository.ts [hunk 6]
```diff
@@ -790,9 +789,9 @@ export class MessageRepository {
     // Configure ephemeral messages
     conversationService.messageTimer.setConversationLevelTimer(conversation.id, conversation.messageTimer());
 
-    const sendOptions: Parameters<typeof conversationService.send>[0] = groupId
+    const sendOptions: Parameters<typeof conversationService.send>[0] = isMLSConversation(conversation)
       ? {
-          groupId,
+          groupId: conversation.groupId,
```
Comment (atomrc) [inline src/script/conversation/MessageRepository.ts:794]: 👍

File: src/script/storage/StorageSchemata.ts [hunk 7]
```diff
@@ -414,6 +416,17 @@ export class StorageSchemata {
...
         version: 19,
       },
+      {
+        schema: {},
+        upgrade: (transaction: Transaction) =>
+          transaction
+            .table(StorageSchemata.OBJECT_STORE.CONVERSATIONS)
+            .toCollection()
+            .modify((conversation: ConversationRecord) => {
+              conversation.initial_protocol = conversation.protocol;
```
Comment (atomrc) [inline src/script/storage/StorageSchemata.ts:426]: good catch 👌

File: src/script/components/MessagesList/Message/ProtocolUpdateMessage/ProtocolUpdateMessage.tsx (hunk 1 above)
Comment (PatrykBuniX) [reply src/script/components/MessagesList/Message/ProtocolUpdateMessage/ProtocolUpdateMessage.tsx:59]: `ProtocolUpdateMessage` (component) renders `SystemMessageBase` (component), `SystemMessageBase` depends on `ProtocolUpdateMessage` (entity). It actually renders any kind of `SystemMessage` (entity) - this is the prop type `SystemMessageBase` requires. Not sure i can see cyclic dependency here 🤔

Comment (atomrc) [reply src/script/components/MessagesList/Message/ProtocolUpdateMessage/ProtocolUpdateMessage.tsx:59]: Ah sorry, I confused `SystemMessageBase` with `SystemMessage`, nvm, you can ignore this comment

File: src/script/components/MessagesList/Message/SystemMessage/SystemMessageBase.tsx (hunk 2 above)
Comment (PatrykBuniX) [reply src/script/components/MessagesList/Message/SystemMessage/SystemMessageBase.tsx:44]: Noted, will think about that 👌

File: src/script/conversation/ConversationRepository.ts (hunk 5 above)
Comment (PatrykBuniX) [reply src/script/conversation/ConversationRepository.ts]: Makes sense 👍

**Review by atomrc — APPROVED** [review]
LGTM. 

Quick word of advice: in general we should name files/directory as nouns, not actions. 
Instead of `finaliseMigration` it should be `migrationFinalizer`.

**Pull request merged**